Applications/Piano/PianoWidget.cpp

```diff
@@ -227,6 +229,60 @@ void PianoWidget::keyup_event(GKeyEvent& event)
     update();
 }
 
+void PianoWidget::mousedown_event(GMouseEvent&)
+{
+    m_mouse_pressed = true;
+
+    if (m_mouse_over == -1)
+        return;
+
+    update_keys();
+    update();
+}
+
+void PianoWidget::mouseup_event(GMouseEvent&)
+{
+    m_mouse_pressed = false;
+
+    if (m_mouse_over == -1)
+        return;
+
+    update_keys();
+    update();
+}
+
+void PianoWidget::mousemove_event(GMouseEvent& event)
+{
+    if (!m_mouse_pressed)
+        return;
+
+    auto rx = event.x() - x();
+    auto ry = event.y() - y();
+
+    int mouse_was_over = m_mouse_over;
+
+    m_mouse_over = -1;
+
+    for (auto& kr : m_key_rects) {
```

Comment:
Why is this two separate loops? Can't we check both white and black keys in the same loop?

Reply:
The reason I do it this way is because the black keys overlap the white keys - so if you have your mouse over the right hand side of a black key, if we check the keys in order, we'll end up picking the white key "under" the black key, which is incorrect.

Reply:
No longer relevant.